platform/security/filter/security-filter-csrf/src/main/java/org/codice/ddf/security/filter/csrf/CsrfFilter.java

```diff
@@ -100,44 +122,101 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
     HttpServletRequest httpRequest = (HttpServletRequest) request;
     HttpServletResponse httpResponse = (HttpServletResponse) response;
     String targetContextPath = httpRequest.getRequestURI();
+    String requestMethod = httpRequest.getMethod();
+    String userAgentHeader = httpRequest.getHeader(HttpHeaders.USER_AGENT);
 
     // Begin CSRF checks if request is accessing a Cross-Site protected context
-    if (protectedContexts.stream().anyMatch(targetContextPath::startsWith)) {
-
-      String sourceOrigin = httpRequest.getHeader(ORIGIN_HEADER);
-      String sourceReferer = httpRequest.getHeader(REFERER_HEADER);
-      String csrfHeader = httpRequest.getHeader(CSRF_HEADER);
-
-      // Reject if no origin or referer header is present on the request
-      if (sourceOrigin == null && sourceReferer == null) {
-        respondForbidden(
-            httpResponse,
-            "Cross-site check failure: Incoming request did not have an Origin or Referer header.");
-        return;
-      }
-
-      // Reject if neither the Origin nor Referer header are a trusted authority
-      if (!isTrustedAuthority(sourceOrigin) && !isTrustedAuthority(sourceReferer)) {
-        respondForbidden(
-            httpResponse,
-            "Cross-site check failure: Neither the Origin nor Referer header matched a system internal or external name.");
-        return;
-      }
+    if (protectedContexts.stream().anyMatch(targetContextPath::startsWith)
+        && doBrowserProtectionFilter(httpRequest, httpResponse, targetContextPath, requestMethod)) {
+      return;
+    }
 
-      // Check for presence of anti-CSRF header
-      // WebSockets API does not allow for custom headers, authority check is sufficient
-      if (!targetContextPath.startsWith(WEBSOCKET_CONTEXT) && csrfHeader == null) {
-        respondForbidden(
-            httpResponse,
-            "Cross-site check failure: Request did not have required X-Requested-With header.");
-        return;
-      }
+    // Execute CSRF check if user is accessing /services
+    if (targetContextPath.startsWith(SERVICE_CONTEXT)
+        && doSystemProtectionFilter(
+            httpResponse, targetContextPath, requestMethod, userAgentHeader)) {
+      return;
     }
 
     // All checks passed
     chain.doFilter(request, response);
   }
 
+  private boolean doSystemProtectionFilter(
+      HttpServletResponse httpResponse,
+      String targetContextPath,
+      String requestMethod,
+      String userAgentHeader) {
+
+    String whitelistKey = getMultivaluedMapKey(SYSTEM_PROTECTION_WHITELIST, targetContextPath);
+    String whitelistUserAgent =
+        USER_AGENT_WHITELIST
+            .stream()
+            .filter(regex -> Pattern.matches(regex, targetContextPath))
```

Comment:
❓ Should this be using the `userAgentHeader` instead of the `targetContextPath` here?